sdk/servicebus/service-bus/src/track2/serviceBusReceiverClient.ts

```diff
@@ -0,0 +1,526 @@
+
+import { ServiceBusMessage, ReceiveMode } from "../serviceBusMessage";
+import { ClientEntityContext } from "../clientEntityContext";
+import { generate_uuid } from "rhea-promise";
+import { ClientType } from "../client";
+import { InternalSessionReceiver, InternalReceiver } from "../internalReceivers";
+import {
+  MessageHandlers,
+  MessageIterator,
+  Session,
+  ContextType,
+  ContextWithSettlement,
+  QueueAuth,
+  SubscriptionAuth,
+  isSession,
+  MessageAndContext,
+  Message,
+  ReceiveBatchOptions,
+  IterateMessagesOptions,
+  StreamMessagesOptions
+} from "./models";
+import { createConnectionContext, convertToInternalReceiveMode } from "./constructorHelpers";
+import { RuleDescription, CorrelationFilter } from "../core/managementClient";
+import { ServiceBusClientOptions } from "../old/serviceBusClient";
+import { ConnectionContext } from "../connectionContext";
+
+/**
+ *A receiver client that handles sessions, including renewing the session lock.
+ */
+// TODO: could extend NonSessionReceiverClient...?
+export interface SessionReceiver<LockModeT extends "peekLock" | "receiveAndDelete"> {
+  /**
+   * Streams messages to message handlers.
+   * @param handler A handler that gets called for messages and errors.
+   * @param options Options for streamMessages.
+   */
+  streamMessages(
+    handlers: MessageHandlers<ContextType<LockModeT>>,
+    options?: StreamMessagesOptions
+  ): void;
+  /**
+   * Returns an iterator that can be used to receive messages from Service Bus.
+   * @param options Options for iterateMessages.
+   */
+  iterateMessages(options?: IterateMessagesOptions): MessageIterator<ContextType<LockModeT>>;
+
+  /**
+   * Receives, at most, `maxMessages` worth of messages.
+   * @param maxMessages The maximum number of messages to accept.
+   * @param maxWaitTimeInSeconds The maximum time to wait, in seconds, for messages to arrive.
+   * @param options Options for receiveBatch.
+   */
+  receiveBatch(
+    maxMessages: number,
+    maxWaitTimeInSeconds?: number,
+    options?: ReceiveBatchOptions
+  ): Promise<Message[]>;
+  /**
+   * Renews the lock on the session.
+   */
+  renewSessionLock(): Promise<Date>;
+  /**
+   * Closes the client.
+   */
+  close(): Promise<void>;
+  /**
+   * Methods related to service bus diagnostics.
+   */
+  diagnostics: {
+    /**
+     * Peek within a queue or subscription.
+     * @param maxMessageCount The maximum number of messages to retrieve.
+     */
+    peek(maxMessageCount?: number): Promise<Message[]>;
+    /**
+     * Peek within a queue or subscription, starting with a specific sequence number.
+     * NOTE: this method does not respect message locks or increment delivery count
+     * for messages.
+     * @param fromSequenceNumber The sequence number to start peeking from (inclusive).
+     * @param maxMessageCount The maximum number of messages to retrieve.
+     */
+    peekBySequenceNumber(fromSequenceNumber: Long, maxMessageCount?: number): Promise<Message[]>;
```

Comment:
I see these have a maxMessageCount...but no maxWaitTime. How long will it actually wait?

Reply:
Look, asking these kinds of questions will only get these issues fixed. (I will check and expose)

Reply:
Interestingly enough we have two kinds of timeouts that might apply here:

1. The core-amqp timeout
2. The other kind of timeout we usually mean in our messaging libraries - return when 'n' messages arrive _or_ 'x' timeout expires.

The first one is 1 minute and we aren't currently exposing it. I'm not actually sure it's one we want to expose in the first place - we should talk about what we did for eventhubs since you did all sorts of timeout work in there.

The second one is probably the one you were talking about and it appears that's not _anywhere_ in the underlying code. So I'll need to add it. 

I'll file an issue to track it.